### PR TITLE
drivers: stm32_rng: Add hardware rng support

### PR DESCRIPTION
**Pull request by GseoC**
Add the hardware RNG support with the device tree driver model.

By default, hardware RNG is used in OP-TEE for STM32MP13 platforms. Non-secure world should use the RNG PTA to generate random numbers.

On STM32MP15 platforms, default implementation remains software RNG. Hardware RNG is used for seeding the RNG but is released post boot.

(Of course, not for 3.20 :) )

### CODE REVIEW

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 1]
```diff
@@ -160,6 +164,11 @@ CFG_STM32_UART ?= y
 CFG_STPMIC1 ?= y
 CFG_TZC400 ?= y
 
+CFG_WITH_SOFTWARE_PRNG ?= n
+ifeq ($(CFG_WITH_SOFTWARE_PRNG),y)
+$(call force,CFG_STM32_RNG,y,Mandated by CFG_WITH_SOFTWARE_PRNG)
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/plat-stm32mp1/conf.mk:169]: I'll change to Required-by as you suggest.

For the second part, wouldn't enforcing CFG_STM32_RNG for STM32MP1x platforms fo the same?

File: core/drivers/stm32_rng.c [hunk 2]
```diff
@@ -222,64 +226,94 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
...
 
-		assert(stm32_rng->clock);
+	if (stm32_rng->rstctrl &&
+	    rstctrl_deassert_to(stm32_rng->rstctrl, RNG_RESET_TIMEOUT_US))
+		goto err;
 
-		DMSG("RNG init");
+	clk_disable(stm32_rng->clock);
+
+	if (IS_ENABLED(CFG_STM32MP15)) {
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_rng.c:507]: Indeed, thank

File: core/drivers/stm32_rng.c [hunk 3]
```diff
@@ -80,19 +80,18 @@ static vaddr_t get_base(void)
...
-static void conceal_seed_error(vaddr_t rng_base)
+static void conceal_seed_error(void)
 {
-	if (io_read32(rng_base + RNG_SR) & (RNG_SR_SECS | RNG_SR_SEIS)) {
-		size_t i = 0;
+	vaddr_t rng_base = get_base();
+	size_t i = 0;
 
-		io_mask32(rng_base + RNG_SR, 0, RNG_SR_SEIS);
+	io_mask32(rng_base + RNG_SR, 0, RNG_SR_SEIS);
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_rng.c:175]: Yes 👍🏻

File: core/arch/arm/plat-stm32mp1/conf.mk (hunk 1 above)
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/plat-stm32mp1/conf.mk:169]: Ok i got what you had in mind, there is no need to force CFG_STM32_RNG if software RNG is configured.

File: core/drivers/stm32_rng.c [hunk 4]
```diff
@@ -161,8 +164,8 @@ TEE_Result stm32_rng_read(uint8_t *out, size_t size)
 		return TEE_ERROR_NOT_SUPPORTED;
 	}
 
-	gate_rng(true, stm32_rng);
-	rng_base = io_pa_or_va(&stm32_rng->base, 1);
+	clk_enable(stm32_rng->clock);
+	rng_base = get_base();
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_rng.c:314]: Ok

**Comment by GseoC** [comment]
Pushed with suggestions applied

File: core/drivers/stm32_rng.c [hunk 5]
```diff
@@ -380,25 +380,23 @@ void plat_rng_init(void)
...
-		 * not taken into account. CONFIGLOCK bit must also be unset but
-		 * it is not handled at the moment.
+		 * not taken into account.
 		 */
-		io_write32(rng_base + RNG_CR,
-			   stm32_rng->pm_cr | RNG_CR_CONDRST);
+		io_write32(base + RNG_CR, stm32_rng->pm_cr | RNG_CR_CONDRST);
 
-		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CONDRST,
+		io_clrsetbits32(base + RNG_CR, RNG_CR_CONDRST,
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c:396]: minor: fits in a single line.

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 6]
```diff
@@ -165,8 +165,8 @@ CFG_STPMIC1 ?= y
 CFG_TZC400 ?= y
 
 CFG_WITH_SOFTWARE_PRNG ?= n
-ifeq ($(CFG_WITH_SOFTWARE_PRNG),y)
-$(call force,CFG_STM32_RNG,y,Mandated by CFG_WITH_SOFTWARE_PRNG)
+ifeq ($(CFG_WITH_SOFTWARE_PRNG),n)
```
Comment (etienne-lms) [inline core/arch/arm/plat-stm32mp1/conf.mk:168]: `ifneq ($(CFG_WITH_SOFTWARE_PRNG),y)`

File: core/drivers/stm32_rng.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_rng.c:314]: This function uses `rng_base` only at 1 place. I think it makes sense to remove this local variable.
This is not a strong opinion, ok as-is if you feel more confortable this way.

File: core/drivers/stm32_rng.c (hunk 5 above)
Comment (GseoC) [reply core/drivers/stm32_rng.c]: Ack

**Comment by GseoC** [comment]
Force-pushed the fixes concerned by your comments

File: core/drivers/stm32_rng.c [hunk 7]
```diff
@@ -219,67 +372,171 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
...
 
-		DMSG("RNG init");
-	}
+	res = clk_enable(stm32_rng->clock);
+	if (res)
+		goto err;
+
+	if (stm32_rng->rstctrl &&
+	    rstctrl_assert_to(stm32_rng->rstctrl, RNG_RESET_TIMEOUT_US))
+		goto err;
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c:492]: set `res = TEE_ERROR_GENERIC;` and disable the clock.
ditto at line 496.

File: core/drivers/stm32_rng.c [hunk 8]
```diff
@@ -283,6 +286,10 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	    rstctrl_deassert_to(stm32_rng->rstctrl, RNG_RESET_TIMEOUT_US))
 		goto err;
 
+	res = init_rng();
+	if (res)
+		goto err;
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c:500]: need to disable the clock, for consistency.

File: core/drivers/stm32_rng.c [hunk 9]
```diff
@@ -140,7 +255,22 @@ static TEE_Result init_rng(void)
...
-	io_setbits32(rng_base + RNG_CR, RNG_CR_RNGEN | RNG_CR_CED);
+	if (stm32_rng->ddata->has_cond_reset) {
+		uint32_t clock_div = stm32_rng_clock_freq_restrain();
+
+		/* Update configuration fields */
+		io_clrsetbits32(rng_base + RNG_CR, RNG_NIST_CONFIG_MASK,
+				RNG_NIST_CONFIG_B | RNG_CR_CONDRST |
+				RNG_CR_CED);
+		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CLKDIV,
+				(clock_div << RNG_CR_CLKDIV_SHIFT));
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c:278]: can remove inner parenthese.

File: core/drivers/stm32_rng.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_rng.c:314]: I prefer to keep it that way for consistency

File: core/drivers/stm32_rng.c (hunk 7 above)
Comment (GseoC) [reply core/drivers/stm32_rng.c:492]: Good spot

File: core/drivers/stm32_rng.c (hunk 8 above)
Comment (GseoC) [reply core/drivers/stm32_rng.c:500]: Yes

File: core/drivers/stm32_rng.c (hunk 9 above)
Comment (GseoC) [reply core/drivers/stm32_rng.c:278]: Ack

**Comment by GseoC** [comment]
Updated with 3 more fix patches

File: core/drivers/stm32_rng.c [hunk 10]
```diff
@@ -367,6 +369,62 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
...
+	assert(stm32_rng);
+
+	res = clk_enable(stm32_rng->clock);
+	if (res)
+		return res;
+
+	if (op == PM_OP_RESUME)
+		res = stm32_rng_pm_resume();
+	else
+		res = stm32_rng_pm_suspend();
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c:426]: Look at how you implemented bsec PM (d6df31b03ea5cd8f84b8a9c99f8559f859bc61d3), maybe do the same here:

```c
static TEE_Result stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
			const struct pm_callback_handle *pm_handle __unused)
{
	static uint32_t pm_cr = 0;
	TEE_Result res = TEE_ERROR_GENERIC;

	assert(stm32_rng);

	res = clk_enable(stm32_rng->clock);
	if (res)
		return res;
	
	if (op == PM_OP_SUSPEND)
		pm_cr = io_read32(get_base() + RNG_CR);
	else
		res = stm32_rng_pm_resume(pm_cr);

	clk_disable(stm32_rng->clock);

	return res;
}

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"plat-stm32mp1: enable hardware rng for STM32MP13",
"drivers: stm32_rng: cleanup and change sample read timeout to 10ms",
"drivers: stm32_rng: call conceal_seed_error() on error cases",
"drivers: stm32_rng: support conditional reset",
"drivers: stm32_rng: support clock error detection",
"drivers: stm32_rng: implement late seed error detection",
"drivers: stm32_rng: override plat_rng_init() if hardware RNG",
and "drivers: stm32_rng: make stm32_rng_read() static".

Suggestion for commit "drivers: stm32_rng: register to dt_driver" even if implementaqtion LGTM. Commit message should also state RNG goes throug a reset cycle when DT provides a reset controller reference.

Suggestion for commit  "drivers: stm32_rng: pm callbacks" but the patch LGTM as is.

For "commit drivers: stm32_rng: keep rng enable from initialization", the patch should also remove `refcount` from `struct stm32_rng_instance`.

File: core/drivers/stm32_rng.c [hunk 11]
```diff
@@ -219,67 +372,179 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
...
+	if (stm32_rng->release_post_boot)
+		stm32mp_register_non_secure_periph_iomem(stm32_rng->base.pa);
+	else
+		stm32mp_register_secure_periph_iomem(stm32_rng->base.pa);
+
+	register_pm_core_service_cb(stm32_rng_pm, &stm32_rng, "rng-service");
+
 	return TEE_SUCCESS;
+
+err:
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_rng.c:521]: Ack, I find the naming clearer the other way around but as it is already implemented that way in TAMP driver, I'll apply your suggestion

File: core/drivers/stm32_rng.c (hunk 10 above)
Comment (GseoC) [reply core/drivers/stm32_rng.c:426]: Ok with your suggestion for consistency.

BTW, I way missing assert (op == PM_OP_SUSPEND || op == PM_OP_RESUME), I'm adding it 👍🏻

**Comment by GseoC** [comment]
> Suggestion for commit "drivers: stm32_rng: register to dt_driver" even if implementaqtion LGTM. Commit message should also state RNG goes throug a reset cycle when DT provides a reset controller reference.

This statement is already present :)

**Comment by GseoC** [comment]
Force-pushed with tags applied on concerned commits + fixes squashed + suggestions applied

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"drivers: stm32_rng: keep rng enable from initialization",
"drivers: stm32_rng: register to dt_driver",
and "drivers: stm32_rng: pm callbacks".

**Comment by GseoC** [comment]
Tags applied, thanks :)

**Pull request merged**